### ISSUE TITLE
[BUG] Wrong attention gradient when Context Parallelism is on

### ISSUE THREAD

**Issue by i-love-megatron** [NVIDIA/Megatron-LM#1465]
**Describe the bug**

- bug location

  When training with context parallelism, the gradient of `core_attention` is wrong.  In another word, the dQ, dK, and dV ourput from the backward function is wrong.  To be more specific, this attention backward function outputs wrong results: `TransformerEngine/transformer_engine/pytorch/attention.py: AttnFuncWithCPAndKVP2P.backward()`.

- how to trigger

  When context parallelism is on and uses this P2P attention module, this bug will be triggered.

- bug behavior

  Given the exactly same parameter and data input for a single card training and a training with context parallelism, the activation gradient of core attention (dQ, dK, dV) do not match in value.  The relative error of these three tensor can be as large as 1.2.  The single card version is double checked with tesnor parallelism and is correct.  

  This will cause the following activation gradient to be wrong in the `linear_qkv` module, and its activation gradient will be used to calculate its parameter gradient.  This error will also accumulate and pollute the gradient calculation of following layers.  The wrong parameter gradient will then accumulate with different micro batches, thus making the `main_grad` for weight update to be wrong, and their relative error will be as large as 2.3.

  So this bug will make the training with context parallelism totally wrong.

**To Reproduce**

Train the GPT model with context parallelism.

1. According to https://github.com/NVIDIA/Megatron-LM/issues/1151, set environment variable `NVTE_BATCH_MHA_P2P_COMM` to 1 to enable P2P communication.

2. Then run `pretrain_gpt.py` with `--context-parallel-size` > 1.

3. Detailed training command:

   ```bash
   CUDA_DEVICE_MAX_CONNECTIONS=1 \
   OMP_NUM_THREADS=1 \
   NVTE_BATCH_MHA_P2P_COMM=1 \
   torchrun \
   --nproc_per_node 2 \
   --nnodes 1 \
   --node_rank 0 \
   --master_addr localhost \
   --master_port 6000 pretrain_gpt.py \
   --tensor-model-parallel-size 1 \
   --pipeline-model-parallel-size 1 \
   --context-parallel-size 2 \  # as long as cp_size > 1
   --num-layers 24 \
   --hidden-size 1024 \
   --num-attention-heads 16 \
   --seq-length 1024 \
   --max-position-embeddings 1024 \
   --micro-batch-size 4 \
   --global-batch-size 16 \
   --lr 0.00015 \
   --train-iters 1 \
   --lr-decay-iters 320000 \
   --lr-decay-style cosine \
   --min-lr 1.0e-5 \
   --weight-decay 1e-2 \
   --lr-warmup-fraction .01 \
   --clip-grad 1.0 \
   --use-mcore-models \
   --attention-softmax-in-fp32 \
   --attention-dropout 0.0 \
   --hidden-dropout 0.0 \
   --bf16 \
   --data-path /workspace/dataset/wikitext/wikitext_text_document \
   --vocab-file /workspace/dataset/gpt2-vocab.json \
   --merge-file /workspace/dataset/gpt2-merges.txt \
   --split 949,50,1 \
   --log-interval 100 \
   --save-interval 10000 \
   --eval-interval 1000 \
   --eval-iters 0 \
   --save /workspace/checkpoints \
   --load /workspace/checkpoints \
   --no-gradient-accumulation-fusion \
   --transformer-impl transformer_engine
   ```

**Expected behavior**

The gradient of attention should be the same as single card when context parallelism is on. 

When it comes to bf16, the relative error of the correctly calculated tensor should be around 5e-3, but the current implementation produces wrong results whose relative error is more than 1 when comparing with correct results.

**Stack trace/logs**

If applicable, add the stack trace or logs from the time of the error.

**Environment (please complete the following information):**
 - Megatron-LM commit ID: [`fdfcef87`](https://github.com/NVIDIA/Megatron-LM/commit/fdfcef87004205eca69ea462b375578185ea01f6)
 - PyTorch version: 2.5.1+cu121
 - CUDA version: 12.1
 - NCCL version: 2.21.5
 - TransformerEngine: [`754d2a0`](https://github.com/NVIDIA/TransformerEngine/commit/754d2a0023131f9e82309f87fe9fc182d1835da8) (current newest commit in `stable` branch)

**Proposed fix**
Check the `AttnFuncWithCPAndKVP2P.backward()` and fix the bug in it.

**Additional context**
Please fix it soon.  This is very serious bug.

**Comment by i-love-megatron** [comment]
Solved in issue https://github.com/NVIDIA/TransformerEngine/issues/1557 and PR https://github.com/NVIDIA/TransformerEngine/pull/1539
